### PR TITLE
Update FHIR Patient Endpoint to Support Standard API Processing

### LINKED ISSUE #3510: `fhir/Patient` FHIR endpoint not working
`fhir/Patient` FHIR endpoint not working with following error:
```[Wed May 13 08:52:41.019313 2020] [php7:error] [pid 619] [client 172.20.0.1:39168] PHP Fatal error:  Uncaught Error: Call to undefined method OpenEMR\\Services\\PatientService::setPid() in /var/www/localhost/htdocs/openemr/src/Services/FHIR/FhirPatientService.php:131\nStack trace:\n#0 /var/www/localhost/htdocs/openemr/src/RestControllers/FHIR/FhirPatientRestController.php(31): OpenEMR\\Services\\FHIR\\FhirPatientService->setId(NULL)\n#1 /var/www/localhost/htdocs/openemr/_rest_routes.inc.php(370): OpenEMR\\RestControllers\\FHIR\\FhirPatientRestController->__construct(NULL)\n#2 [internal function]: {closure}()\n#3 /var/www/localhost/htdocs/openemr/src/Common/Http/HttpRestRouteHandler.php(32): call_user_func_array(Object(Closure), Array)\n#4 /var/www/localhost/htdocs/openemr/apis/dispatch.php(151): OpenEMR\\Common\\Http\\HttpRestRouteHandler::dispatch(Array, '/fhir/Patient', 'GET')\n#5 {main}\n  thrown in /var/www/localhost/htdocs/openemr/src/Services/FHIR/FhirPatientService.php on line 131```

### PR DESCRIPTION
**Pull request by dixonwhitmire**
PR to address FHIR Patient Processing Issues
This is a rebase of #3513 - I'll remove the "WIP" once I've reviewed the rebase.

Fixes #3510
- Added FhirServiceBase to define required FHIR processing operations - data mapping, query processing, etc
- Updated FhirPatientService to support FhirServiceBase expectations
- Implemented Fhir Patient query parameters per spec based on current OpenEMR support
- Updated tests and adjusted api/controller code and tests as necessary
- Updated PatientService to support new UUID registry and “standard” OpenEMR API processing.

Note: FHIR API Responses have not been updated to adhere to the FHIR R4 specification. Updating the API to support FHIR R4 is larger GSOC effort. From what I could tell the existing implementation wasn't uniform or complete. This PR just "unbreaks" the patient FHIR and OpenEMR API endpoints and integrates them with the "standard" OpenEMR API response.

### CODE REVIEW

**Comment by tywrenn** [comment]
You probably didn't notice this yet your email isn't set in your git client. Please run the following command in git, replacing the email area with your GitHub account email address:
`git config --global user.email "MY_NAME@example.com"`

File: src/Common/Uuid/UuidRegistry.php [hunk 1]
```diff
@@ -120,4 +121,22 @@ public function tableNeedsUuidCreation()
...
+    }
+
+    /**
+     * Converts a UUID string to a bytes representation
+     * @return the UUID bytes value
+     */
+    public static function uuidToBytes($uuidString)
+    {
+        return Uuid::fromString($uuidString)->getBytes();
+    }
```
Comment (bradymiller) [inline src/Common/Uuid/UuidRegistry.php:141]: 2DV

File: src/Services/FHIR/FhirPatientService.php [hunk 2]
```diff
@@ -2,167 +2,382 @@
 
 namespace OpenEMR\Services\FHIR;
 
+use Ramsey\Uuid\Uuid;
```
Comment (bradymiller) [inline src/Services/FHIR/FhirPatientService.php:5]: Not seeing where this class is being used in this script

**Comment by dixonwhitmire** [comment]
Hi @yashrajbothra and @sjpadgett - this is the rebased PR for the FHIR patient service integration.

File: src/Services/FHIR/FhirPatientService.php (hunk 2 above)
Comment (dixonwhitmire) [reply src/Services/FHIR/FhirPatientService.php:5]: Good catch! I can remove the import/use statement.

File: src/Services/FHIR/FhirResourcesService.php [hunk 3]
```diff
@@ -16,6 +16,8 @@
 use OpenEMR\FHIR\R4\FHIRResource\FHIRBundle\FHIRBundleLink;
 use OpenEMR\FHIR\R4\PHPFHIRResponseParser;
 
+require_once('_rest_config.php');
```
Comment (bradymiller) [inline src/Services/FHIR/FhirResourcesService.php:19]: are we sure this relative path will always work. seems better to use an absolute path here:
```
require_once(__DIR__ . '/../../../_rest_config.php');
```

Comment (dixonwhitmire) [reply src/Services/FHIR/FhirResourcesService.php:19]: You know I had a heck of a time getting that to work. I will add the DIR and path info in and retest

**Comment by bradymiller** [comment]
hi @dixonwhitmire , Code looks good to me (except for the super minor path issue). If testing well for you, then I thing we should bring this in to keep the fhir momentum going :) . @sjpadgett and @yashrajbothra , just let us know if note issues.

**Comment by sjpadgett** [comment]
Okay with me!

File: _rest_routes.inc.php [hunk 4]
```diff
@@ -355,13 +355,28 @@
...
+    },
+    "PUT /fhir/Patient/:id" => function ($pid) {
+        RestConfig::authorization_check("patients", "demo");
+        $data = (array)(json_decode(file_get_contents("php://input"), true));
+        return (new FhirPatientRestController())->put($id, $data);
+    },
+    "PATCH /fhir/Patient/:pid" => function ($id) {
+        RestConfig::authorization_check("patients", "demo");
+        $data = (array)(json_decode(file_get_contents("php://input"), true));
+        return (new FhirPatientRestController())->put($id, $data);
```
Comment (bradymiller) [inline _rest_routes.inc.php:371]: quick likely dumb question, why do the PUT/PATCH above both not use `:id` and `$id`?

File: src/RestControllers/FHIR/FhirPatientRestController.php [hunk 5]
```diff
@@ -9,119 +9,109 @@
...
 namespace OpenEMR\RestControllers\FHIR;
 
 use OpenEMR\Services\FHIR\FhirResourcesService;
 use OpenEMR\Services\FHIR\FhirPatientService;
 use OpenEMR\Services\FHIR\FhirValidationService;
 use OpenEMR\RestControllers\RestControllerHelper;
 use OpenEMR\FHIR\R4\FHIRResource\FHIRBundle\FHIRBundleEntry;
+use OpenEMR\Validators\ProcessingResult;
+
+require_once('_rest_config.php');
```
Comment (bradymiller) [inline src/RestControllers/FHIR/FhirPatientRestController.php:22]: another place to consider an absolute path

File: src/Services/FHIR/FhirResourcesService.php (hunk 3 above)
Comment (sjpadgett) [reply src/Services/FHIR/FhirResourcesService.php:19]: Should be able to put this in the OpenEMR namespace and path your way to a use statement, true?

Comment (sjpadgett) [reply src/Services/FHIR/FhirResourcesService.php:19]: @dixonwhitmire just leave this and i'll move and refactor this.

File: _rest_routes.inc.php (hunk 4 above)
Comment (dixonwhitmire) [reply _rest_routes.inc.php:371]: Hey @bradymiller they should be consistent! I'll update the routes.

**Comment by dixonwhitmire** [comment]
@bradymiller @sjpadgett @yashrajbothra - Thanks for the review! I did make another update to address the "variable" name mismatch for the FHIR endpoint "REST routes".

While the FHIR endpoints aren't quite there in regards to adhering to the R4 spec, what this PR has accomplished is how to integrate the OpenEMR REST and FHIR APIs with the "standard API process" which is documented in the [API README](https://github.com/openemr/openemr/blob/master/API_README.md). Hopefully APIs are easier to develop and maintain since they utilize a standard response format. 

FHIR endpoints are implemented atop of the REST API, so the FHIR endpoints and services will need to consume the standard response from the REST API before constructing a FHIR response such as an OperationOutcome, Bundle, etc.

Yash and I did a bit of a [review](https://docs.google.com/document/d/1zAmXE4YNF0ut2lBagGlQXoMEEI8EoVpes7tc9gxeh0M/edit) on his design doc to discuss options for implementation. The code I've added in may need to be updated based on what is learned when the implementation begins.

Thanks!

**Comment by bradymiller** [comment]
hi @dixonwhitmire , All sounds good and will bring this in now. Thanks for the awesome updates! -brady

**Comment by bradymiller** [comment]
Time for a fhir cheer by the regulars!

2DV

**Comment by sjpadgett** [comment]
Thanks for the hard work @dixonwhitmire ! We're lucky to have ya on board.

**Comment by dixonwhitmire** [comment]
@sjpadgett - my pleasure!

**Pull request merged**